napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,122 @@
+    ) -> Future[dict]:
+        """This should only be called from the main thread.
+
+        Creates a new task and adds it to the _layers_to_task dict. Cancels
+        all tasks currently pending for that layer.
+
+        Submitting multiple layers at one generates multiple requests (stored in a dict),
+        but only ONE task.
+
+        A slice submitted with multiple layers will only cancel if the same tuple of layers
+        is reubmitted. If a second slice is requested with only one layer from the first
+        slice, it will not be cancelled.
+        """
+        # Cancel any tasks that are slicing a subset of the layers
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        if existing_task := self._find_existing_task(layers):
+            LOGGER.debug('Cancelling task for %s', layers)
+            existing_task.cancel()
+
+        # Not all layer types will initially be asynchronously sliceable.
+        # The following logic gives us a way to handle those in the short
+        # term as we develop, and also in the long term if there are cases
+        # when we want to perform sync slicing anyway.
+        requests = {}
+        for layer in layers:
+            # TODO: temporary to prove a point (which currently fails)
+            # import weakref
+            # assert isinstance(layer, weakref.ReferenceType) # this fails
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # create task for slicing of each request/layer
+        task = self._executor.submit(self._slice_layers, requests)
+
+        # construct dict of layers to layer task
+        self._layers_to_task[tuple(requests.keys())] = task
+
+        # once everything is complete, release the block
+        task.add_done_callback(self._on_slice_done)
+
+        return task
+
+    def shutdown(self) -> None:
+        """This should be called from the main thread when this is no longer needed."""
+        # TODO: This makes this class instance null. Consider a context manager
+        #       to handle instances of this class.
+        self._executor.shutdown()
+
+    def _slice_layers(self, requests: Dict) -> Dict:
+        """This can be called from the main or slicing thread.
+        Iterates throught a dictionary of request objects and call the slice
+        on each individual layer."""
+        return {
+            layer: layer._get_slice(request)
+            for layer, request in requests.items()
+        }
+
+    def _on_slice_done(self, task: Future[Dict]) -> None:
+        """This can be called from the main or slicing thread.
+        Release the thread.
+        This is the "done_callback" which is added to each task.
+        """
+        if not self._try_to_remove_task(task):
```

Comment:
I wonder if instead of explicitly removing the tasks we could use [`WeakValueDictionary`](https://docs.python.org/3/library/weakref.html#weakref.WeakValueDictionary). I think the executor should be the only thing storing a strong reference to the future and it should dispose of that once the done callbacks complete.

Reply:
I still think this might be a good idea, but can be done in follow up PR.